concert/ext/tangoservers/qa.py

```diff
@@ -0,0 +1,318 @@
+
+    @DebugIt()
+    @command()
+    async def update_darks(self) -> None:
+        await self._process_stream(self._update_ff_projections("_dark", self._receiver.subscribe()))
+
+    @DebugIt()
+    @command()
+    async def update_flats(self) -> None:
+        await self._process_stream(self._update_ff_projections("_flat", self._receiver.subscribe()))
+
+    @DebugIt()
+    @command(
+        dtype_in=(float,),
+        doc_in="number of markers, wait window, check window, sigma, decision threshold"
+    )
+    async def estimate_center_of_rotation(self, args: Tuple[float]) -> None:
+        """Estimates the center of rotation"""
+        num_markers, wait_window, check_window = int(args[0]), int(args[1]), int(args[2])
+        sigma, err_threshold = args[3], args[4]
+        self.info_stream("%s starting estimation using markers=%d, wait_time=%d, sigma=%f",
+                         self.__class__.__name__, num_markers, wait_window, sigma)
+        await self._process_stream(self._estimate_from_markers(self._receiver.subscribe(),
+                                                               num_markers=num_markers,
+                                                               wait_window=wait_window,
+                                                               check_window=check_window,
+                                                               sigma=sigma,
+                                                               err_threshold=err_threshold))
+
+    @staticmethod
+    def opt_func(angle_x: np.float64, center_p1: np.float64, radius_p2: np.float64,
+                 phase_p3: np.float64) -> np.float64:
+        """Defines the model function for nonlinear polynomial fit"""
+        return center_p1 + radius_p2 * np.cos(angle_x + phase_p3)
+
+    async def _estimate_from_markers(self, producer: AsyncIterator[ArrayLike], num_markers: int,
+                                     wait_window: int, check_window: int, sigma: float,
+                                     err_threshold: float) -> None:
+        """
+        Estimates the center of rotation with marker tracking and nonlinear polynomial fit
+        """
+        projection_count = 0
+        marker_centroids: List[ArrayLike] = []
+        estm_rot_axis: List[float] = []
+        event_triggered = False
+        num_intensity_classes = 3
+        ffc = FlatCorrect(dark=self._dark, flat=self._flat, absorptivity=False)
+        # Derive Gaussian kernel size from standard deviation with formula: 2 * ceil(3 * sigma) + 1.
+        lpf = GaussianFilter(kernel_size=2 * np.ceil(3 * sigma).astype(int) + 1, sigma=sigma)
+        #lpf = MedianFilter(kernel_size=3)
+        # Process all projections for potential marker centroids but run curve-fit conditionally.
+        async for projection in lpf(ffc(producer)):
+            thresh: ArrayLike = skf.threshold_multiotsu(projection, classes=num_intensity_classes)
```

Comment:
Why multi-otsu? You have background and one "class" with multiple instances, i.e. more labels, but not more "histogram blobs".

Reply:
Or do you intend to segment the markers and the sample in one shot? If so, how to choose which threshold belongs to what? Anyway, as a proof of concept for the actual fitting it's OK, but we need a way more robust way to distinguish the markers from the sample.

Reply:
Adapted segmentation approach.